keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,1767 @@
+errors as described in goals above.
+
+
+### Solutions and Gaps Without Runtime Assisted Mounts
+
+- A pod using a microvm runtime (like Kata) can mount PVs backed by block
+devices as a raw block device to ensure the file system mount is managed within
+the pod sandbox environment. However this approach has the following shortcomings:
+  - Every workload pod that mounts PVs backed by block devices needs to mount
+  the file system (on the raw block PV). This requires modifications to the
+  workload container logic. Therefore this is not a general purpose solution.
+  - File system management operations like online expansion of the file system
+  and reporting of FS stats cannot be performed at an infrastructural level if
+  the file system mounts are managed by workload pods.
+- A pod using a microvm runtime (like Kata) may use a filesystem (e.g. virtiofs)
+to project the file-system on-disk (and mounted on the host) to the guest
+environment (as detailed in the diagram below). While this solves several
+use-cases, it is not desired due to the following factors:
+  - Native file-system features: The workload may wish to use specific features
+  and system calls supported by an on-disk file system (e.g. ext4/xfs) mounted
+  on the host. However, plumbing some of these features/system calls/file
+  controls through the intermediate “projection” file system (e.g. virtio-fs)
+  and it’s dependent framework (e.g. FUSE) may require extra work and therefore
+  not supported until implemented and tested. Some examples of this
+  incompatibility are: [open_by_handle_at](https://gitlab.com/virtio-fs/qemu/-/issues/10), [xattr](https://gitlab.com/virtio-fs/qemu/-/issues/15), [F_SETLKW](https://gitlab.com/virtio-fs/qemu/-/issues/9) and fscrypt
+  - Safety and isolation: A compromised or malicious workload may try to exploit
+  vulnerabilities in the file system of persistent volumes. Due to the inherently
+  complex nature of file system modules running in kernel mode, they present a
+  larger surface of attack relative to simpler/lower level modules in the
+  disk/block stack. Therefore, it is safer to mount the file system within an
+  isolated sandbox environment (like guest VM) rather than in the host
+  environment as the blast radius of a file system kernel mode panic/DoS will be
+  isolated to the sandbox rather than affecting the entire host. A recent example
+  of a vulnerability in virtio-fs: https://nvd.nist.gov/vuln/detail/CVE-2020-10717
+  - Performance: As pointed out [here](https://archive.fosdem.org/2020/schedule/event/vai_virtio_fs/attachments/slides/3666/export/events/attachments/vai_virtio_fs/slides/3666/virtio_fs_A_Shared_File_System_for_Virtual_Machines_FOSDEM.pdf) [slide 6], in
+  a microvm environment, a block interface (such as virtio-blk) provides a
+  faster path to sharing data between a guest and host relative to a file-system
+  interface.
+
+![Current Mount Workflow](images/CurrentMounts.png)
+
+## Proposal
```

Comment:
As mentioned above (or below?), this KEP leaks some Kubernetes and Kata details into CSI, which was container orchestrator and container runtime agnostic. It's a huge change in CSI design.

(At the same time, I don't see much opportunity to do it outside of CSI, unless we allow a piece of a CSI driver to run inside the micro VM)